code/modules/telecomms/telecomms.dm

```diff
@@ -0,0 +1,222 @@
+/obj/machinery/telecomms_machine/attack_hand(var/mob/user)
+	if(src.stat & (EMPED|NOPOWER|BROKEN))
+		return
+	var/dat = "<font face = \"Courier\"><HEAD><TITLE>[src.name]</TITLE></HEAD><center><H3>[src.name] Access</H3></center>"
+	dat += "<br>Power Status: <a href='?src=\ref[src];toggle=1'>[src.on ? "On" : "Off"]</a>"
+	dat += "<br>Operating frequencies: "
+	if(src.channels.len)
+		for(var/T in src.channels)
+			dat += "[format_frequency(T)] GHz<a href='?src=\ref[src];delete=[T]'>\[X\]</a>; "
+	else
+		dat += "none; "
+	dat += "<a href='?src=\ref[src];addfreq=1'>add new</a>."
+	dat += "</font>"
+	user << browse(dat, "window=tcommachine;size=520x500;can_resize=0")
+
+/obj/machinery/telecomms_machine/Topic(href, href_list)
+	if(src.stat & (EMPED|NOPOWER|BROKEN))
+		return
+	if(href_list["toggle"])
+		src.on = !src.on
+	else if(href_list["addfreq"])
+		var/newfreq = input(usr, "Specify a new frequency to filter (GHz). Decimals assigned automatically.") as null|num
+		if(newfreq)
+			if(findtext(num2text(newfreq), "."))
+				newfreq *= 10 // shift the decimal one place
+			if(!(newfreq in src.channels) && newfreq < 10000)
+				src.channels.Add(newfreq)
+	else if(href_list["delete"])
+		var/x = text2num(href_list["delete"])
+		src.channels.Remove(x)
+	src.attack_hand(usr)
+
+/obj/machinery/telecomms_machine/attackby(var/obj/item/I, var/mob/user)
+	if(default_deconstruction_screwdriver(user, I))
+		updateUsrDialog()
+		return
+	if(default_deconstruction_crowbar(user, I))
+		return
+	if(default_part_replacement(user, I))
+		return
+
+	if(istype(I, /obj/item/stack/nanopaste))
+		if(integrity < 100)
+			var/obj/item/stack/nanopaste/N = I
+			if(N.use(1))
+				stat &= ~BROKEN
+				integrity = between(0, integrity + 35, 100)
+				user.visible_message("<span class='notice'>\The [user] applies \the [N] to \the [src].</span>", "You apply \the [N] to \the [src], repairing [integrity == 100 ? "all" : "some"] of the damage.")
+			return
+		else
+			user << "<span class='notice'>This machine is already in perfect condition.</span>"
+
+/obj/machinery/telecomms_machine/common
+	channels = list(PUB_FREQ)
+
+/obj/machinery/telecomms_machine/allinone
+	channels = list(PUB_FREQ, SCI_FREQ, COMM_FREQ, MED_FREQ, ENG_FREQ, SEC_FREQ, ERT_FREQ, DTH_FREQ, SYND_FREQ, SUP_FREQ, SRV_FREQ, AI_FREQ)
+
+/obj/machinery/telecomms_machine/proc/processSignal(var/mob/living/source, var/mobName, var/sayVerb, var/message, var/freq, var/datum/language/speaking)
+	if(!(freq in src.channels))
+		return 0
+
+	if(!src.isUsable())
+		return 0
+
+	broadcastMessage(source, mobName, sayVerb, message, freq, speaking, 0)
+	return 1
+
+/obj/machinery/telecomms_machine/proc/isUsable()
+	if(!src.on)
+		return 0
+	if(stat & (EMPED|NOPOWER|BROKEN))
+		return 0
+	return 1
+
+/proc/broadcastMessage(var/mob/living/source, var/mobName, var/sayVerb, var/message, var/freq, var/datum/language/speaking, var/messageFlags)
+	var/list/radios = list()
+
+	for(var/obj/item/device/radio/R in world) // AAAAAAAAAAAAAA
```

Comment:
On one hand, thanks for trying to help, on an another, do you really think I don't realize this?